packages/flutter/lib/src/rendering/object.dart

```diff
@@ -2594,14 +2594,14 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
     } else {
       assert(() {
         if (debugPrintMarkNeedsPaintStacks) {
-          debugPrintStack(label: 'markNeedsPaint() called for $this (root of render tree)');
+          debugPrintStack(label: 'markNeedsPaint() called for $this');
         }
         return true;
       }());
-      // If we're the root of the render tree (probably a RenderView),
-      // then we have to paint ourselves, since nobody else can paint
-      // us. We don't add ourselves to _nodesNeedingPaint in this
-      // case, because the root is always told to paint regardless.
+      // This case does not happen in normal trees.
```

Comment:
Are there cases in which this can still happen? If not, we should just delete this branch then and assert that the case can't happen.

Reply:
Maybe a Flutter user who, for some weird reason, choose to have a RenderObject that is neither a repaint boundary nor having a parent RenderObject? I mainly do this for compatibility reasons, but if you like I can remove and assert.

Reply:
Ok, the tree doesn't have to be rooted at a RenderView and therefore the root may not be a relayoutboundary. Therefore, this code is still needed. Let's update the comment to something like this:

> If we are the root of the render tree (and not a repaint boundary) then we have to paint ourselves, since nobody else can paint us. We don't add ourselves to _nodesNeedingPaint in this case, because the root is always told to paint regardless.
>
> Typical trees rooted at a RenderView do not go through this code path because RenderViews are repaint boundaries.

Reply:
done